ruby/ql/src/queries/security/cwe-116/IncompleteMultiCharacterSanitization.ql

```diff
@@ -0,0 +1,27 @@
+/**
+ * @name Incomplete multi-character sanitization
+ * @description A sanitizer that removes a sequence of characters may reintroduce the dangerous sequence.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.8
+ * @precision high
+ * @id rb/incomplete-multi-character-sanitization
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-020
+ *       external/cwe/cwe-080
+ *       external/cwe/cwe-116
+ */
+
+import ruby
+import codeql.ruby.frameworks.core.String
+import codeql.ruby.DataFlow
+import codeql.ruby.security.IncompleteMultiCharacterSanitizationQuery
+import codeql.ruby.security.IncompleteMultiCharacterSanitizationSpecific as Specific
+
+from
+  StringSubstitutionCall replace, Specific::EmptyReplaceRegExpTerm dangerous, string prefix,
```

Comment:
You could reduce the number of imports a bit by declaring more the types like this:
```suggestion
  DataFlow::Node replace, RegExpTerm dangerous, string prefix,
```
The `hasResult` predicate takes care of restricting the types.

Alternatively you can remove the entire `from .. where .. select` and write:

```
query predicate problems = hasResult/4;
```
An perhaps even better, you can rename `hasResult` to `problems` and make it a `query` predicate in the QLL file. That way all you have to do in the QL file is `import codeql.ruby.security.IncompleteMultiCharacterSanitizationQuery`

Reply:
The predicate should be problems(replace, message, dangerous, prefix). Just drop the kind parameter and insert the message string with placeholder as the second parameter.